In [1]:
import pandas as pd

# Parameters

# Downloading the data

In [4]:
dataframe_path = "dataset.csv"

In [5]:
df = pd.read_csv(dataframe_path, index_col=0)

Date  distance_fire_stations  distance_rivers  distance_roads  \
0  2015-11-19            13287.682266      7211.102551     1250.000000   
1  2003-02-17             8721.381771      2358.495283      250.000000   
2  2012-02-26            10796.411441         0.000000     2015.564437   
3  2004-11-10             8253.787010       559.016994        0.000000   
4  2003-03-19             9905.806378         0.000000     1903.943276   

   distance_powerlines  cropland  forest_deciduous_broad  \
0         30196.233209       0.0                     0.0   
1         13768.169813       0.0                     0.0   
2          6254.998002       0.0                     0.0   
3         37350.535471       0.0                     0.0   
4          6427.480066       0.0                     0.0   

   forest_deciduous_needle  forest_evergreen_broad  forest_evergreen_needle  \
0                      0.0                1.000000                      0.0   
1                      0.0                0.416667                      0.0   
2                      0.0                0.666667                      0.0   
3                      0.0                0.000000                      0.0   
4                      0.0                0.750000                      0.0   

   ...  avg_wind_angle  avg_rel_hum  avg_soil  sum_prec    forest  \
0  ...      225.773605    56.837185  0.297854  0.360376  1.000000   
1  ...      209.708847    61.120739  0.264534  0.020176  0.833333   
2  ...       76.341278    63.017559  0.208871  0.025395  1.000000   
3  ...       68.557823    64.673866  0.156506  0.000000  0.000000   
4  ...      316.951508    56.103680  0.208831  0.119717  0.916667   

   vegetation_class  Year  max_max_temp  yearly_avg_temp  ignition  
0            forest  2015     62.552337        14.994683         1  
1            forest  2003     60.787457        15.053698         1  
2            forest  2012     63.420256        15.001883         1  
3           wetland  2004     60.394119        14.850611         1  
4            forest  2003     69.570496              NaN         1  

[5 rows x 42 columns]

# Structure of the dataframe and Task


- Each row consists of an ignition or non-ignition point with the given features associated. The features were chosen as potentially influencing ignition.


- The last column named `ignition` says if the point was a real ignition point (meaning that it occurred historically), in that case the value is `1`. Otherwise, when the value is `0`, it means that it is a 'non-ignition point'.


- The columns `cropland` to `wetland` gives the ratio of each of the vegetation classes under which the ignition or non-ignition point lies. The sum of these ratios should be equal to 1. For more information refer to this website: https://lcviewer.vito.be/


- The temperatures should be in degrees celsius `(°C)`.


- The weather data come from different sources and they might have different units.

| Column name | Definition | Unit |
|--------|-----------|--------|
| `ignition`   | Target column| Boolean: {1,0} |
| `distance_{feature}` | Distance to nearest feature  |  Meters (m) |
| vegetation class: from `cropland` to `wetland`  |  Ratio of each of the vegetation classes  under which the ignition or non-ignition point lies    |  No unit (between 0 and 1)  |
| `aspect`  |  Orientation of the slope    |  Degrees (°)  |
| `elevation`  |  elevation value    |  Meters  |
| `slope`  |  Slope value    |  Degrees (°)  |
| `pop_dens`  |  Population density value    |  Persons per km2  |
| `max_temp`  |  Maximum temperature of the day    |  Degrees celsius (°C)  |
| `avg_temp`  |  Average temperature of the day   |  Degrees celsius (°C)  |
| `max_wind_vel`  |  Maximum wind velocity of the day    |  Meters per second (m/s)  |
| `avg_wind_angle`  |  Average angle of the vector wind over the day    |  Degrees (°)  |
| `avg_rel_hum`  |  Average relative humidity over the day    |  %  |
| `avg_soil`  |  Average soil moisture of the day    |  m3/m3  |
| `sum_prec`  |  Cumulative rainfall precipitation of the day    |  Millimeters (mm)  |
| `yearly_avg_temp`  |  Average temperature over the year    |  Degrees celsius (°C)  |
| `anom_{feature}`  |  Standardized anomaly of weather for the given day over the last 30 years. When the anomaly is positive, it means that the feature value is greater that the 30-year average    |  No unity |
| `forest`  |  Sum of all the columns where the names start with `forest`   |  No unit  |
| `vegetation_class`  |  Vegetation with the max occurrence in the vicinity of the ignition/non-ignition point    |  Without unit  |
| `Year`  |  Year of ignition    |  Without unit  |
| `max_max_temp`  |  Missing information    |  Missing information  |



# Main steps of a data science project

```mermaid
graph TD
    A[Start] --> B[Identify Project Objectives]
    B --> C[Identify Required Data]
    C --> D[Load the Data]
    D --> E[Explore and Understand Data]
    E --> F[Preprocess the Data]
    F --> G[Feature Selection]
    G --> H[Split Data into Training and Test Sets]
    H --> I[Select and Train Algorithm]
    I --> J[Evaluate the Model]
    J --> K{Model Performance Acceptable?}
    K -- Yes --> L[Deploy / Communicate Results]
    K -- No --> F

Before starting any data science project, it essential to clearly define the objectives of the project. The project objectives guide all subsequent steps, including what algorithms and what data is to be used.

Once the objectives have been established, the next step is to source the data that is required to carry out the project objectives.

With the data in-hand, the focus of the project shifts to understand the data. This phase of the project deals with uncovers data patterns and potential issues. Following exploring the data, the data must be cleaned and processed. The preprocessing of the data is one of the most important steps in any data science project and often determining the overall success of the project. Data processing can be considered a "make or break" for any project.

Once the data is preprocessed, the features and labels must be selected and the dataset split into training and test sets.

Next, select an appropriate algorithm based on the project objectives. It is typically considered good practice to select multiple algorithms and evaluate their performance. Note that each algorithm must suit the task at hand and must be adapated as needed. For example, the of use logistic regression for classification tasks over linear regression.

With the data prepared and algorithms selected, the models are then trained and evaluated. If the performance is acceptable, the results should be presented and the model may proceed to deployment. However, if the model performance is unsatisfactory, one must go back to preprocessing and feature selection to investigate why the model underperformed. The goal of this step should be to identify whether the poor performance stems from insufficient data cleaning, suboptimal features or a bad model choice. Iterate through these steps until the model meets the project's success criteria.

# Carrying Out the Classification Project

## Problem Definition and Objective Identification

The goal of this project is to create a classification algorithm based on a number of features from a given dataset, the classifier will be able to predict whether a given set of features correspond to an ignition or not. 

The dataset is a multidimensional space having 42 columns, one of which is the target label, ignition. Ignition is either 1 or 0. Meaning that this task is a **binary classification task**.

The next step is to understand the data and try find correlations in the data. It is expected that in such a large dataset, not all features are going to have strong correlations and therefore, may be considered for removal. 

## Understanding and Preprocessing the Data

In [6]:
# Getting the information on the dataset such as datatypes
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 22035 entries, 0 to 22034
Data columns (total 42 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Date                     22035 non-null  object 
 1   distance_fire_stations   22030 non-null  float64
 2   distance_rivers          22030 non-null  float64
 3   distance_roads           22030 non-null  float64
 4   distance_powerlines      22030 non-null  float64
 5   cropland                 22030 non-null  float64
 6   forest_deciduous_broad   22030 non-null  float64
 7   forest_deciduous_needle  22030 non-null  float64
 8   forest_evergreen_broad   22030 non-null  float64
 9   forest_evergreen_needle  22030 non-null  float64
 10  forest_mixed             22030 non-null  float64
 11  forest_unknown           22030 non-null  float64
 12  herbaceous_vegetation    22030 non-null  float64
 13  moss_lichen              22030 non-null  float64
 14  shrubland                22

In [7]:
# Describing the dataset to see the style of data that is being dealt with
df.describe()

distance_fire_stations  distance_rivers  distance_roads  \
count            22030.000000     22030.000000    22030.000000   
mean             23646.387792      5966.777537     5152.597702   
std              19248.657525      7515.660146     6924.754655   
min                  0.000000         0.000000        0.000000   
25%               9568.829605       707.106781      250.000000   
50%              18494.931738      2610.076627     1581.138830   
75%              32310.988843      8384.323013     7941.190087   
max             110474.261708     43784.986011    40094.419811   

       distance_powerlines      cropland  forest_deciduous_broad  \
count         22030.000000  22030.000000            22030.000000   
mean          30127.951951      0.006415                0.013557   
std           30099.446768      0.057848                0.088239   
min               0.000000      0.000000                0.000000   
25%            8265.137627      0.000000                0.000000   
50%           20846.161854      0.000000                0.000000   
75%           41340.053217      0.000000                0.000000   
max          159274.919557      1.000000                1.000000   

       forest_deciduous_needle  forest_evergreen_broad  \
count                  22030.0            22030.000000   
mean                       0.0                0.326108   
std                        0.0                0.433120   
min                        0.0                0.000000   
25%                        0.0                0.000000   
50%                        0.0                0.000000   
75%                        0.0                0.916667   
max                        0.0                1.000000   

       forest_evergreen_needle  forest_mixed  ...      avg_temp  \
count             22030.000000       22030.0  ...  22035.000000   
mean              -4584.657286           0.0  ...    158.859675   
std               20915.699938           0.0  ...    135.801507   
min             -100000.000000           0.0  ...     -4.109107   
25%                   0.000000           0.0  ...      7.687609   
50%                   0.000000           0.0  ...    276.348625   
75%                   0.000000           0.0  ...    280.721741   
max                   0.000000           0.0  ...    292.551632   

       avg_wind_angle   avg_rel_hum      avg_soil      sum_prec        forest  \
count    22035.000000  22035.000000  22035.000000  22035.000000  22035.000000   
mean       197.707583     76.109929      0.285717      1.662872      0.426235   
std         83.336828      8.027890      0.078390      3.230198      0.455856   
min          0.792745     40.705662      0.050758     -0.000007      0.000000   
25%        133.581253     70.623573      0.233063      0.000000      0.000000   
50%        203.072937     76.005028      0.287844      0.154361      0.166667   
75%        257.033798     81.626057      0.346534      1.854040      1.000000   
max        358.530182     98.558968      0.501387     40.440075      1.000000   

               Year  max_max_temp  yearly_avg_temp      ignition  
count  22035.000000  22035.000000     15204.000000  22035.000000  
mean    2011.278784     59.230929        14.999755      0.150669  
std        5.693506      7.983432         0.100675      0.357734  
min     2001.000000     34.109174        14.597322      0.000000  
25%     2006.000000     53.440945        14.931427      0.000000  
50%     2012.000000     58.614709        14.999861      0.000000  
75%     2016.000000     64.858667        15.068628      0.000000  
max     2021.000000     90.376239        15.369208      1.000000  

[8 rows x 40 columns]

In [8]:
import numpy as np

In [9]:
# Selecting the data with datatypes flaot64 and int64 to see if there are linear correlations with ignition
df_correlation = df.select_dtypes(include=['float64', 'int64'])
df_correlation.corr()['ignition'].sort_values()

avg_rel_hum               -0.269664
distance_roads            -0.268259
water                     -0.254303
distance_fire_stations    -0.228656
distance_rivers           -0.218113
distance_powerlines       -0.205647
anom_avg_soil             -0.192976
anom_avg_rel_hum          -0.156313
avg_soil                  -0.136861
max_wind_vel              -0.110371
anom_sum_prec             -0.108719
sum_prec                  -0.089841
cropland                  -0.022591
sprarse_vegetation        -0.006609
aspect                    -0.004460
Year                       0.007185
yearly_avg_temp            0.012401
forest_evergreen_needle    0.012867
avg_temp                   0.016031
shrubland                  0.026196
anom_max_wind_vel          0.028030
anom_avg_temp              0.040474
avg_wind_angle             0.052949
elevation                  0.052994
herbaceous_vegetation      0.061074
wetland                    0.076226
forest_evergreen_broad     0.091934
forest_deciduous_broad     0

Since the goal of this task is to classify whether certain data corresponds to an ignition point, the correlation between data and ignition was checked. It was noted that forest_deciduous_needle, forest_mixed and moss_lichen had no correlation with ignition, indicating that these most likely had no variation in data. Therefore, these 3 datafields are not important and can therefore be dropped.

Looking at the information on the dataset, yearly_avg_temp was observed to have approximately 6800 NaN datapoints. Yearly average temperature could influence whether a point is classified as an ignition point however, the data needs to be looked into further before dropping this data. The options for this column are to either impute the data should it be valuable or drop the column entirely.

Moreover, there appears to be overlapping data such as Date and Year, naturally the Year column can be dropped and Date used. It may be worth converting Date into Month, Year and Day of the year as wildfires could be seasonal.


In [10]:
# Dropping NaN correlated rows
df.drop(columns=['forest_deciduous_needle', 'forest_mixed', 'moss_lichen'], inplace=True)

# Handling Date and converting to standalone columns
df['Date'] = pd.to_datetime(df['Date'])
df['Month'] = df['Date'].dt.month
df['DayofYear'] = df['Date'].dt.dayofyear
df.drop(columns=['Year'], inplace=True)
df['Year'] = df['Date'].dt.year
df.drop(columns=['Date'], inplace=True)

Since yearly_avg_temp is a yearly average, the data can be grouped by year and using the valid data, the NaN values can be converted to the mean of that year. This preserves possibly impactful data rather than just removing it.

In [11]:
# Imputing yearly_avg_temp by filling the NaN values with the mean yearly_avg_temp of that year from the valid data
df['yearly_avg_temp'] = df.groupby('Year')['yearly_avg_temp'].transform(lambda x: x.fillna(x.mean()))
print("Remaining NaNs in yearly_avg_temp :", df['yearly_avg_temp'].isna().sum())

Remaining NaNs in yearly_avg_temp : 0


From the information on the dataset, it was noted that columns from cropland to wetland gives the ratio of vegetation class. Moreover, the sum of these ratios must be 1. Therefore, this must be checked.

In [12]:
# Listing the vegetation columns
veg_columns = [
    'cropland',
    'forest_deciduous_broad',
    'forest_evergreen_broad',
    'forest_evergreen_needle',
    'forest_unknown',
    'herbaceous_vegetation',
    'shrubland',
    'sprarse_vegetation',
    'urban',
    'water',
    'wetland'
]

# Summing the ratios if the vegetation class
row_sums = df[veg_columns].sum(axis=1)
# Creating a mask that returns True if the sum is not correct. Note that the condition is not == 1 as to account for computational precision
invalid_sum_mask = ~row_sums.between(0.999, 1.001)

# Checking what classes are contributing to the bad summation by collecting the values that are outside the range [0,1]
invalid_values = (df[veg_columns] < 0) | (df[veg_columns] > 1)
invalid_by_class = invalid_values.sum().sort_values(ascending=False)
invalid_by_class

forest_evergreen_needle    1010
forest_deciduous_broad        0
cropland                      0
forest_evergreen_broad        0
forest_unknown                0
herbaceous_vegetation         0
shrubland                     0
sprarse_vegetation            0
urban                         0
water                         0
wetland                       0
dtype: int64

Only entries from forest_evergreen_needle are outside the range, let us inspect what the bad data looks like

In [13]:
# Selecting the invalid datapoints from forest_evergreen_needle
invalid_forest_evergreen_needle = df.loc[
    (df['forest_evergreen_needle'] < 0) | (df['forest_evergreen_needle'] > 1),
    'forest_evergreen_needle'
]

# Seeing what the erraneous values are
invalid_value_counts = invalid_forest_evergreen_needle.value_counts()
invalid_value_counts

forest_evergreen_needle
-100000.0    1010
Name: count, dtype: int64

The erraneous values are all identical and an extreme value. The first check is to set the extreme value to 0, if the sum is good after setting the extreme values to 0 then the error can be considered fixed. If not, these rows should be dropped.

In [14]:
# Creating a mask that returns true if the value is outside the range [0,1]
invalid_mask = (df['forest_evergreen_needle'] < 0) | (df['forest_evergreen_needle'] > 1)

# Setting the extreme values to zero
df.loc[invalid_mask, 'forest_evergreen_needle'] = 0
# Recalculating the sum and checking if the == 1 condition is true
row_sums = df[veg_columns].sum(axis=1)
invalid_sum_mask = ~row_sums.between(0.999, 1.001)
num_invalid_rows = invalid_sum_mask.sum()
print('The number of remaning invalid rows are:', num_invalid_rows)

The number of remaning invalid rows are: 5


From the above it can be seen that setting the extreme value to 0 corrected the sum. Now only 5 datapoints are invalid. 

In [15]:
# Selecting the rows that still have an invalid sum
df_invalid_sum = df.loc[invalid_sum_mask, veg_columns]

df_invalid_sum

cropland  forest_deciduous_broad  forest_evergreen_broad  \
8031        NaN                     NaN                     NaN   
10398       NaN                     NaN                     NaN   
14704       NaN                     NaN                     NaN   
17444       NaN                     NaN                     NaN   
20679       NaN                     NaN                     NaN   

       forest_evergreen_needle  forest_unknown  herbaceous_vegetation  \
8031                       NaN             NaN                    NaN   
10398                      NaN             NaN                    NaN   
14704                      NaN             NaN                    NaN   
17444                      NaN             NaN                    NaN   
20679                      NaN             NaN                    NaN   

       shrubland  sprarse_vegetation  urban  water  wetland  
8031         NaN                 NaN    NaN    NaN      NaN  
10398        NaN                 NaN    NaN    NaN      NaN  
14704        NaN                 NaN    NaN    NaN      NaN  
17444        NaN                 NaN    NaN    NaN      NaN  
20679        NaN                 NaN    NaN    NaN      NaN

It can be seen that the 5 rows with invalid sums have NaN entries and can be dropped. However, prior to dropping the NaN rows, it is better to first go through all the data as dropping NaN rows early in the data processing, may drop useful data.

From the information of the dataset, the weather data comes from multiple sources and may have inconsistent units. Therefore, the weather data specifically will be checked to ensure consistency

In [16]:
# Checking temperature fields to ensure consistency
df[['max_temp', 'avg_temp', 'yearly_avg_temp', 'max_max_temp','anom_max_temp', 'anom_avg_temp']].describe()

max_temp      avg_temp  yearly_avg_temp  max_max_temp  \
count  22035.000000  22035.000000     22035.000000  22035.000000   
mean      15.128294    158.859675        14.999730     59.230929   
std        4.435240    135.801507         0.083652      7.983432   
min        1.171763     -4.109107        14.597322     34.109174   
25%       11.911636      7.687609        14.963919     53.440945   
50%       14.785950    276.348625        14.999554     58.614709   
75%       18.254815    280.721741        15.034747     64.858667   
max       32.431244    292.551632        15.369208     90.376239   

       anom_max_temp  anom_avg_temp  
count   22035.000000   22035.000000  
mean        2.044202       1.941795  
std         3.026061       2.559182  
min        -7.159943      -9.482144  
25%        -0.098090       0.202694  
50%         1.768574       1.928117  
75%         3.918903       3.628255  
max        16.340687      13.051181

From the above description, it can be seen that max_temp and yearly_avg_temp are in degrees celsius however, avg_temp has a mixture of units. The maximum value is 292.55 - which is impossible should it be in celsius or fahrenheit, therefore it must be in Kelvin. However, it was also noted that the minimum is -4.1 which is standard for celsius.

From this description it can be seen that avg_temp has a mixture of data in Celsius and data in Kelvin.

Furthermore, it can be seen that max_max_temp falls within a standard fahrenheit range. Therefore, it is undoubtadely in Fahrenheit and can be converted using a simple formula.

In [17]:
# Assuming that the max_temp values are true and therefore, it can be assumed that any temperature over 35 is definitely in Kelvin
kelvin_mask = df['avg_temp'] > 35
# Selecting the datapoints that are captured by the mask and converting them to degrees celsius
df.loc[kelvin_mask, 'avg_temp'] = df.loc[kelvin_mask, 'avg_temp'] - 273.15

# Converting max_max_temp into degrees celsius
df['max_max_temp'] = (df['max_max_temp'] - 32) / 1.8

In [18]:
# Rechecking to ensure consistency
df[['max_temp', 'avg_temp', 'yearly_avg_temp', 'max_max_temp','anom_max_temp', 'anom_avg_temp']].describe()

max_temp      avg_temp  yearly_avg_temp  max_max_temp  \
count  22035.000000  22035.000000     22035.000000  22035.000000   
mean      15.128294      7.316278        14.999730     15.128294   
std        4.435240      3.285425         0.083652      4.435240   
min        1.171763     -4.109107        14.597322      1.171763   
25%       11.911636      5.025002        14.963919     11.911636   
50%       14.785950      7.156210        14.999554     14.785950   
75%       18.254815      9.465587        15.034747     18.254815   
max       32.431244     19.401632        15.369208     32.431244   

       anom_max_temp  anom_avg_temp  
count   22035.000000   22035.000000  
mean        2.044202       1.941795  
std         3.026061       2.559182  
min        -7.159943      -9.482144  
25%        -0.098090       0.202694  
50%         1.768574       1.928117  
75%         3.918903       3.628255  
max        16.340687      13.051181

It is now evident that the temperature fields are all in the same units. Moreover, max_max_temp has the same data as max_temp and can therefore be dropped.

In [19]:
df.drop(columns=['max_max_temp'], inplace=True)

In [20]:
# Checking wind speed fields to ensure consistency
df[['max_wind_vel', 'anom_max_wind_vel']].describe()

max_wind_vel  anom_max_wind_vel
count  22035.000000       22035.000000
mean       5.728723           2.202667
std        2.483441           3.320765
min        0.978202          -7.363498
25%        3.858555          -0.172485
50%        5.250791           1.964479
75%        7.180290           4.269049
max       19.703424          22.360945

It can be seen that the wind speed variations are of the same magnitude and are therefore, consistent in units

In [21]:
# Checking precipitation amounts to ensure consistency
df[['sum_prec', 'anom_sum_prec']].describe()

sum_prec  anom_sum_prec
count  22035.000000   22035.000000
mean       1.662872       0.040211
std        3.230198       1.042084
min       -0.000007      -3.103708
25%        0.000000      -0.710898
50%        0.154361      -0.090009
75%        1.854040       0.691713
max       40.440075       4.699601

There is a wide range in the precipitation data, from no rainfall to a maximum of 40.44. Moreover, 75% of the data is within 1.85, indicating that this large 40.44 could be classed as an outlier, indicating an extreme case. Realistically 40.44 mm of rainfall in a day is a lot but plausible with extreme rainfall. Therefore, no adjustments need to be done for the precipitation data.

Moreover, it can be seen that the anom_{features} have non-zero means and non-one standard deviations. These indicating that the standardisation done prior to data loading was done badly or not even standardised. Only anom_sum_prec appears to be standardised well. Therefore, the rest should be restandardised or dropped. First let's inspect all the anom_{features}

In [22]:
# Selectin anom_{features}
anom_cols = [col for col in df.columns if col.startswith('anom_')]
# Describing the anom_{features}
df[anom_cols].describe()

anom_max_temp  anom_max_wind_vel  anom_avg_temp  anom_avg_rel_hum  \
count   22035.000000       22035.000000   22035.000000      22035.000000   
mean        2.044202           2.202667       1.941795          2.231864   
std         3.026061           3.320765       2.559182          2.796936   
min        -7.159943          -7.363498      -9.482144         -9.155835   
25%        -0.098090          -0.172485       0.202694          0.293640   
50%         1.768574           1.964479       1.928117          2.066875   
75%         3.918903           4.269049       3.628255          4.100976   
max        16.340687          22.360945      13.051181         15.005052   

       anom_avg_soil  anom_sum_prec  
count   22035.000000   22035.000000  
mean        0.458071       0.040211  
std         1.651015       1.042084  
min        -6.903322      -3.103708  
25%        -0.727537      -0.710898  
50%         0.439721      -0.090009  
75%         1.647857       0.691713  
max         8.484861       4.699601

In [23]:
# Performing column-wise standardisation on all the anom_{features}
for col in anom_cols:
    mean = df[col].mean()
    std = df[col].std()
    df[col] = (df[col] - mean) / std
df[anom_cols].describe()

anom_max_temp  anom_max_wind_vel  anom_avg_temp  anom_avg_rel_hum  \
count   2.203500e+04       2.203500e+04   2.203500e+04      2.203500e+04   
mean    2.579688e-17       1.305967e-17   8.383985e-18      2.579688e-17   
std     1.000000e+00       1.000000e+00   1.000000e+00      1.000000e+00   
min    -3.041626e+00      -2.880711e+00  -4.463903e+00     -4.071490e+00   
25%    -7.079473e-01      -7.152424e-01  -6.795536e-01     -6.929810e-01   
50%    -9.108472e-02      -7.172687e-02  -5.344783e-03     -5.898941e-02   
75%     6.195189e-01       6.222608e-01   6.589839e-01      6.682710e-01   
max     4.724454e+00       6.070371e+00   4.340991e+00      4.566849e+00   

       anom_avg_soil  anom_sum_prec  
count   2.203500e+04   2.203500e+04  
mean   -3.611563e-17   2.837656e-17  
std     1.000000e+00   1.000000e+00  
min    -4.458708e+00  -3.016953e+00  
25%    -7.181091e-01  -7.207753e-01  
50%    -1.111437e-02  -1.249613e-01  
75%     7.206386e-01   6.251922e-01  
max     4.861731e+00   4.471224e+00

anom_{features} are now properly standardised

In [24]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 22035 entries, 0 to 22034
Data columns (total 39 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   distance_fire_stations   22030 non-null  float64
 1   distance_rivers          22030 non-null  float64
 2   distance_roads           22030 non-null  float64
 3   distance_powerlines      22030 non-null  float64
 4   cropland                 22030 non-null  float64
 5   forest_deciduous_broad   22030 non-null  float64
 6   forest_evergreen_broad   22030 non-null  float64
 7   forest_evergreen_needle  22030 non-null  float64
 8   forest_unknown           22030 non-null  float64
 9   herbaceous_vegetation    22030 non-null  float64
 10  shrubland                22030 non-null  float64
 11  sprarse_vegetation       22030 non-null  float64
 12  urban                    22030 non-null  float64
 13  water                    22030 non-null  float64
 14  wetland                  22

Now only 32 rows contain NaN data and can therefore be dropped as their influence can be assumed negligible since 32 << 22035

In [25]:
# Dropping the rows with NaN entries
df = df.dropna()

From the orginal description of the dataframe, it was noted that the minimum distance values are 0. Although this is possible, it is highly unlikely. For example, how likely is it that an ignition of a wildfire started at a fire station?

The number should be checked and the entries with 0 distance should be processed.

In [26]:
(df[['distance_roads', 'distance_fire_stations', 'distance_powerlines', 'distance_rivers']] == 0).sum()

distance_roads            3103
distance_fire_stations       3
distance_powerlines        157
distance_rivers           1337
dtype: int64

It is likely that an ignition point occured on roads, powerlines and rivers, so these could be true. Although it is very unlikely that an ignition occured at fire station, only 3 ignitions occured at a fire station. Therefore, it is highly rare but possible. 

As a result, nothing will be done to these datapoints as they are not unreasonable.

All that is left is to process the vegetation_class as it is of type object. It can be assumed that there may be errors in the string entries.

In [27]:
# Checking the unique classes in vegetation_class
df['vegetation_class'].unique()

array(['forest', 'wetland', 'herbaceous_vegetation', 'Forestt',
       'shrubland', 'water', 'urban', '$herb$aceous_vegetation'],
      dtype=object)

As expected, there are spelling mistakes these must be arranged. Since, there are only a few these can be arranged simply through a mapping

In [28]:
# Creating a mapping dictionary
veg_mapping = {
    'Forestt': 'forest',
    '$herb$aceous_vegetation': 'herbaceous_vegetation',
    np.nan: 'unknown'
}

# Applying the mapping
df['vegetation_class'] = df['vegetation_class'].replace(veg_mapping)

# Checking the uniqueness again
df['vegetation_class'].unique()

array(['forest', 'wetland', 'herbaceous_vegetation', 'shrubland', 'water',
       'urban'], dtype=object)

To be able to train a classification model, the objects must be one-hot encoded

In [29]:
# One-hot encoding using pandas
df = pd.get_dummies(df, columns=['vegetation_class'], drop_first=True, dtype=int)
df

distance_fire_stations  distance_rivers  distance_roads  \
0                13287.682266      7211.102551     1250.000000   
1                 8721.381771      2358.495283      250.000000   
2                10796.411441         0.000000     2015.564437   
3                 8253.787010       559.016994        0.000000   
4                 9905.806378         0.000000     1903.943276   
...                       ...              ...             ...   
22030            12260.199835      1820.027472      500.000000   
22031             8933.784193      3889.087297      790.569415   
22032            56560.255480      1030.776406     6388.466170   
22033            94191.294715     16839.314119    14637.281168   
22034            41926.274578       707.106781     3092.329219   

       distance_powerlines  cropland  forest_deciduous_broad  \
0             30196.233209       0.0                     0.0   
1             13768.169813       0.0                     0.0   
2              6254.998002       0.0                     0.0   
3             37350.535471       0.0                     0.0   
4              6427.480066       0.0                     0.0   
...                    ...       ...                     ...   
22030         39374.007924       0.0                     0.0   
22031         40380.998006       0.0                     0.0   
22032         23538.532240       0.0                     0.0   
22033         83236.485390       0.0                     0.0   
22034         39957.790229       0.0                     0.0   

       forest_evergreen_broad  forest_evergreen_needle  forest_unknown  \
0                    1.000000                      0.0        0.000000   
1                    0.416667                      0.0        0.416667   
2                    0.666667                      0.0        0.333333   
3                    0.000000                      0.0        0.000000   
4                    0.750000                      0.0        0.166667   
...                       ...                      ...             ...   
22030                0.000000                      0.0        0.000000   
22031                1.000000                      0.0        0.000000   
22032                1.000000                      0.0        0.000000   
22033                0.000000                      0.0        0.000000   
22034                0.000000                      0.0        0.000000   

       herbaceous_vegetation  ...  yearly_avg_temp  ignition  Month  \
0                   0.000000  ...        14.994683         1     11   
1                   0.166667  ...        15.053698         1      2   
2                   0.000000  ...        15.001883         1      2   
3                   0.166667  ...        14.850611         1     11   
4                   0.083333  ...        15.002587         1      3   
...                      ...  ...              ...       ...    ...   
22030               0.000000  ...        14.994649         0      5   
22031               0.000000  ...        15.045942         0     10   
22032               0.000000  ...        14.982935         0      6   
22033               0.000000  ...        14.897028         0      6   
22034               0.500000  ...        15.002066         0      1   

       DayofYear  Year  vegetation_class_herbaceous_vegetation  \
0            323  2015                                       0   
1             48  2003                                       0   
2             57  2012                                       0   
3            315  2004                                       0   
4             78  2003                                       0   
...          ...   ...                                     ...   
22030        137  2013                                       0   
22031        295  2017                                       0   
22032        154  2019                                       0   
22033        157  2007                     

## Feature Selection and Splitting the Data into Train and Test Sets
Prior to splitting the dataset into train and test sets, it is valuable to see the distribution of classes that is, the balance between the number of ignition and non-ignition points. It is important that the train and test sets are statistically representative of the data

In [30]:
from sklearn.model_selection import train_test_split

# Selecting features by dropping the ignition column
X = df.drop(columns='ignition')
# Selecting ignition as the label
y = df['ignition']

print("This is the distribution of non-ignition and ignition points", y.value_counts() * 100 / y.count())

This is the distribution of non-ignition and ignition points ignition
0    84.911148
1    15.088852
Name: count, dtype: float64


There is an imbalance of ignition and non-ignition points therefore, it is required that the test and train sets have that same representation

In [31]:
# Using sklearn's dataset splitter
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42, stratify=y)
print("This is the distribution of non-ignition and ignition points in train set", y_train.value_counts() * 100 / y_train.count())
print("This is the distribution of non-ignition and ignition points in test set", y_test.value_counts() * 100 / y_test.count())

This is the distribution of non-ignition and ignition points in train set ignition
0    84.910806
1    15.089194
Name: count, dtype: float64
This is the distribution of non-ignition and ignition points in test set ignition
0    84.91252
1    15.08748
Name: count, dtype: float64


Using the stratify function, sklearn preserved the data split. The float64 data must now be standardised. Standardisation is done using a scaler fitted to the training data only because in reality the model works on un-seen data which the scaler will not be refit on. The model needs to be able to handle data standardised using its training parameters.

In [32]:
from sklearn.preprocessing import StandardScaler

# Selecting the float columns and non-float columns
float_cols = X_train.select_dtypes(include='float64').columns
non_float_cols = X_train.columns.difference(float_cols)

# Fitting a scaler on the training data only but then applying it to both the train and test data
scaler = StandardScaler()
X_train_scaled_float = scaler.fit_transform(X_train[float_cols])
X_test_scaled_float = scaler.transform(X_test[float_cols])

# Converting back into a DataFrame
X_train_scaled = pd.DataFrame(X_train_scaled_float, columns=float_cols, index=X_train.index)
X_test_scaled = pd.DataFrame(X_test_scaled_float, columns=float_cols, index=X_test.index)

# Concatenating to include the one-hot encoded entries
X_train_final = pd.concat([X_train_scaled, X_train[non_float_cols]], axis=1)
X_test_final = pd.concat([X_test_scaled, X_test[non_float_cols]], axis=1)

# Extracting the values from the DataFrame
X_test_values = X_test_final.values
X_train_values = X_train_final.values

y_test_values = y_test.values
y_train_values = y_train.values

## Algorithm Selection, Training and Evaluation

From the above data distribution and the project objective, the overall task is **stratified binary classification**. 

In terms of algorithm selection, possible candidates are:

1. Logistic regression
2. Support Vector Machines
3. Random Forest
4. Gradient Boosted Classification
5. Neural Networks

It is best practice to always start from the simplest algorithm, in this case logistic regression. Not only does starting from the simplest algorithm provide a solid baseline for comparison, but also avoid using too complex and costly algorithms.

In [33]:
# Importing the required packages
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report

# Defining the model
LR_model = LogisticRegression( class_weight='balanced', random_state=42, max_iter = 5000)
LR_model.fit(X_train_values, y_train_values)

# Predicting on the test set
y_pred = LR_model.predict(X_test_values)

# Computing the performance metrics
print(classification_report(y_test_values, y_pred, digits=3))

              precision    recall  f1-score   support

           0      0.964     0.777     0.860      3737
           1      0.400     0.837     0.541       664

    accuracy                          0.786      4401
   macro avg      0.682     0.807     0.701      4401
weighted avg      0.879     0.786     0.812      4401



In [34]:
# Importing required packages
from sklearn.svm import LinearSVC

# Defining the model
svm_model = LinearSVC(class_weight='balanced', random_state=42)

# Training the model
svm_model.fit(X_train_values, y_train_values)

# Predicting on the test set
y_pred_sgd = svm_model.predict(X_test_values)

print(classification_report(y_test_values, y_pred_sgd, digits=3))

              precision    recall  f1-score   support

           0      0.966     0.771     0.857      3737
           1      0.396     0.846     0.540       664

    accuracy                          0.782      4401
   macro avg      0.681     0.809     0.699      4401
weighted avg      0.880     0.782     0.810      4401



In [35]:
# Importing required packages
from sklearn.ensemble import RandomForestClassifier

# Defining the model
rf_model = RandomForestClassifier(class_weight='balanced', n_estimators=100, random_state=42)

# Training the model
rf_model.fit(X_train_values, y_train_values)

# Predicting on the test set
y_pred_rf = rf_model.predict(X_test_values)

print(classification_report(y_test_values, y_pred_rf, digits=3))

              precision    recall  f1-score   support

           0      0.900     0.980     0.938      3737
           1      0.775     0.384     0.514       664

    accuracy                          0.890      4401
   macro avg      0.837     0.682     0.726      4401
weighted avg      0.881     0.890     0.874      4401



In [36]:
# Importing required packages
from sklearn.ensemble import GradientBoostingClassifier

# Defining the model
gb_model = GradientBoostingClassifier(n_estimators=100, random_state=42)

# Training the model
gb_model.fit(X_train_values, y_train_values)

# Predicting on the test set
y_pred_gb = gb_model.predict(X_test_values)

print(classification_report(y_test_values, y_pred_gb, digits=3))

              precision    recall  f1-score   support

           0      0.906     0.965     0.935      3737
           1      0.691     0.434     0.533       664

    accuracy                          0.885      4401
   macro avg      0.798     0.700     0.734      4401
weighted avg      0.873     0.885     0.874      4401



In [37]:
from sklearn.neural_network import MLPClassifier

# Defining the model
nn_model = MLPClassifier(
    hidden_layer_sizes=(32, 32),  
    activation='relu',            
    solver='adam',                 
    max_iter=5000,                 
    random_state=42,
    learning_rate_init=1e-4
)

# Training the model
nn_model.fit(X_train_values, y_train_values)

# Predicting on the test set
y_pred_nn = nn_model.predict(X_test_values)

print(classification_report(y_test_values, y_pred_nn, digits=3))

              precision    recall  f1-score   support

           0      0.857     0.991     0.919      3737
           1      0.593     0.072     0.129       664

    accuracy                          0.853      4401
   macro avg      0.725     0.532     0.524      4401
weighted avg      0.817     0.853     0.800      4401



### Evaluating the models

To evaluate the models, sklearn's classification report was used. The classification report gives detailed insight into the model's capability to predict both classes. Prior to comparing the models, it is important to understand what each scoring metric represents

- **Precision**: Represents that ratio of true positives and the total number of predicted positives. In other words, of all the times the model predicted a class, how often was it correct
- **Recall**: Represents the ratio of true positives and the total number of actual positives. In other words, of all the actual ignition events, how many were correctly identified.
- **F1 score**: This is a balance between the precision and recall. It provides a measure when both false positives and false negatives are important

Accuracy is avoided as it is not a reliable metric in this case due to the imbalance in classes. Ignition events are much rarer than non-ignition events. A model that predicts no igntion each time would still achieve high accuracy while failing to identify actual ignition points. This would give a false impression of the model quality.

In the context of parametric insurance for wildfire cases, the primary goal is to correctly identify as many ignition points as possible. Missing an ignition (a false negative) could lead to serious consequences, that is not paying out a client when they should have been paid out, while over predicting (a false positive) would lead to paying out clients when not needed. Therefore, model evaluation is primarily focused on the precision and recall of the ignition cases. This meaning that we want high scores for precision and recall for the case the class is 1 (ignition) and would manifest itself in a high f1 score for class 1.


#### Comparison of the Tested Models

As discussed above, 5 models algorithms were tested. Firstly, it is important to note that none of these algorithms were highly optimised as performance was not the main goal of this project.

When comparing the unoptimised models, the logistic regression algorithm worked best with an f1 score of 0.541 followed closely by the SVM with an f1 score of 0.540 and gradient boosting algorithm with an f1 score of 0.533. The reason for the logistic regression algorithm being the highest performing could be due to it being the simplest and therefore requiring little to no tuning. On the other hand, more complex algorithms such as random forest, gradient boost and NNs are capable of much higher performance but require more attentative parameter tuning, therefore their lack of performance can be attributed to a lack of tuning.